crates/llms/src/embeddings/candle/mod.rs

```diff
@@ -0,0 +1,238 @@
+            Err(e) => {
+                tracing::error!("Failed to lock backend: {:?}", e);
+                return Err(super::Error::FailedToCreateEmbedding {
+                    source: "Failed to lock backend".into(),
+                });
+            }
+        };
+
+        Ok(pooled_embeddings)
+    }
+
+    fn size(&self) -> i32 {
+        self.model_cfg.hidden_size
+    }
+}
+
+/// For a given `HuggingFace` repo, download the needed files to create a `CandleEmbedding`.
+pub fn download_hf_artifacts(model_id: &str, revision: Option<&str>) -> Result<PathBuf> {
+    let api = ApiBuilder::new()
+        .with_progress(false)
+        .build()
+        .boxed()
+        .context(FailedToInstantiateEmbeddingModelSnafu)?;
+
+    let repo = if let Some(revision) = revision {
+        Repo::with_revision(model_id.to_string(), RepoType::Model, revision.to_string())
+    } else {
+        Repo::new(model_id.to_string(), RepoType::Model)
+    };
+    let api_repo = api.repo(repo.clone());
+
+    tracing::trace!("Downloading 'config.json' for {}", repo.url());
+    api_repo
+        .get("config.json")
+        .boxed()
+        .context(FailedToInstantiateEmbeddingModelSnafu)?;
+
+    tracing::trace!("Downloading 'tokenizer.json' for {}", repo.url());
+    api_repo
+        .get("tokenizer.json")
+        .boxed()
+        .context(FailedToInstantiateEmbeddingModelSnafu)?;
+
+    tracing::trace!("Downloading 'model.safetensors' for {}", repo.url());
+    let model = if let Ok(p) = api_repo.get("model.safetensors") {
+        p
+    } else {
+        let p = api_repo
+            .get("pytorch_model.bin")
+            .boxed()
+            .context(FailedToInstantiateEmbeddingModelSnafu)?;
+        tracing::warn!("`model.safetensors` not found. Using `pytorch_model.bin` instead. Model loading will be significantly slower.");
+        p
+    };
+    Ok(model
+        .parent()
+        .ok_or("".into())
+        .context(FailedToInstantiateEmbeddingModelSnafu)?
+        .to_path_buf())
+}
+
+/// Create a temporary directory with the provided files softlinked into the base folder (i.e not nested). The files are linked with to names defined in the hashmap, as keys.
+///
+/// Example:
+///
+/// ```rust
+/// use std::collections::HashMap;
+/// use std::path::Path;
+///
+/// let files: HashMap<String, &Path> = vec![
+///    ("model.safetensors".to_string(), Path::new("path/to/model.safetensors")),
+///   ("config.json".to_string(), Path::new("path/to/irrelevant_filename.json")),
+/// ].into_iter().collect();
+///
+/// let temp_dir = link_files_into_tmp_dir(files).unwrap();
+///
+/// ```
+///
+fn link_files_into_tmp_dir(files: HashMap<String, &Path>) -> Result<PathBuf> {
+    let temp_dir = tempdir()
+        .boxed()
+        .context(FailedToInstantiateEmbeddingModelSnafu)?;
+
+    for (name, file) in files {
+        symlink(file, &temp_dir.path().join(name))
```

Comment:
This won't work on Windows

Reply:
Using  `std::os::windows::fs::symlink_file` based on OS `cfg!` has limitations [too](https://doc.rust-lang.org/std/os/windows/fs/fn.symlink_file.html#limitations).